examples/mnist-classification/src/mnist_classification/core.clj

```diff
@@ -0,0 +1,214 @@
+   (layers/convolutional 5 0 1 20)
+   (layers/max-pooling 2 0 2)
+   (layers/dropout 0.9)
+   (layers/relu)
+   (layers/convolutional 5 0 1 50)
+   (layers/max-pooling 2 0 2)
+   (layers/batch-normalization)
+   (layers/linear 1000)
+   (layers/relu :center-loss {:label-indexes {:stream :labels}
+                              :label-inverse-counts {:stream :labels}
+                              :labels {:stream :labels}
+                              :alpha 0.9
+                              :lambda 1e-4})
+   (layers/dropout 0.5)
+   (layers/linear num-classes)
+   (layers/softmax :id :labels)])
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Get Yann LeCun's mnist dataset and save it as folders of folders of png
+;; files. The top level folders are named 'training' and 'test'. The subfolders
+;; are named with class names, and those folders are filled with images of the
+;; appropriate class.
+(defn- ds-data->png
+  "Given data from the original dataset, use think.image to produce png data."
+  [ds-data]
+  (let [data-bytes (byte-array (* image-size image-size))
+        num-pixels (alength data-bytes)
+        retval (image/new-image image/*default-image-impl*
+                                image-size image-size :gray)]
+    (c-for [idx 0 (< idx num-pixels) (inc idx)]
+           (let [[x y] [(mod idx image-size)
+                        (quot idx image-size)]]
+             (aset data-bytes idx
+                   (unchecked-byte (* 255.0
+                                      (+ 0.5 (m/mget ds-data y x)))))))
+    (image/array-> retval data-bytes)))
+
+
+(defn- save-image!
+  "Save a dataset image to disk."
+  [output-dir [idx {:keys [data label]}]]
+  (let [image-path (format "%s/%s/%s.png" output-dir (util/max-index label) idx)]
+    (when-not (.exists (io/file image-path))
+      (io/make-parents image-path)
+      (i/save (ds-data->png data) image-path))
+    nil))
+
+;; These two defonces use helpers from cortex to procure the original dataset.
+(defn- timed-get-dataset
+  [f name]
+  (println "Loading" name "dataset.")
+  (let [start-time (System/currentTimeMillis)
+        ds (f)]
+    (println (format "Done loading %s dataset in %ss"
+                     name (/ (- (System/currentTimeMillis) start-time) 1000.0)))
+    ds))
+
+(defonce training-dataset
+  (timed-get-dataset mnist/training-dataset "mnist training"))
+
+(defonce test-dataset
+  (timed-get-dataset mnist/test-dataset "mnist test"))
+
+
+(def dataset-folder "mnist/")
+
+
+(defonce ensure-images-on-disk!
+  (memoize
+   (fn []
+     (println "Ensuring image data is built, and available on disk.")
+     (dorun (map (partial save-image! (str dataset-folder "training"))
+                 (map-indexed vector training-dataset)))
+     (dorun (map (partial save-image! (str dataset-folder "test"))
+                 (map-indexed vector test-dataset)))
+     :done)))
+
+
+(def max-image-rotation-degrees 25)
```

Comment:
Better to put the magic-number type of defs at the top of the file / section and the defonces following them.